File: generated_code/python/test_unittest_3915.py
import os
import importlib.util
import unittest

# Dynamic loading of the solution module as per guidelines
solution_path = os.environ.get("TEST_SOLUTION_FILE")
if not solution_path:
    raise RuntimeError("TEST_SOLUTION_FILE environment variable not set.")

spec = importlib.util.spec_from_file_location("Solution", solution_path)
sol_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(sol_module)
Solution = sol_module.Solution


class TestMaxProductNoCommonBits(unittest.TestCase):
    def setUp(self):
        self.sol = Solution()

    def test_example_case_with_valid_pairs(self):
        """
        Tests a typical case where valid pairs with no common set bits exist.
        e.g., nums = [4, 8, 3, 5]
        Binary: 4 (0100), 8 (1000), 3 (0011), 5 (0101)
        Valid pairs: (4, 3) -> no common bits, product = 12
                     (8, 3) -> no common bits, product = 24
                     (8, 5) -> no common bits, product = 40
        Max product should be 40.
        """
        nums = [4, 8, 3, 5]
        self.assertEqual(self.sol.maxProduct(nums), 40)

    def test_no_valid_pair_exists(self):
        """
        Tests a case where every number shares at least one bit with every other number,
        meaning no valid pair exists. Should return 0.
        """
        # All numbers have the lowest bit set (odd numbers)
        nums = [1, 3, 5, 7]
        self.assertEqual(self.sol.maxProduct(nums), 0)

    def test_minimum_constraints(self):
        """
        Tests the absolute minimum constraints: length of nums is 2.
        Case A: Valid pair.
        Case B: Invalid pair (shares bits).
        """
        # Case A: 2 (0010) and 5 (0101) -> no common bits, product = 10
        self.assertEqual(self.sol.maxProduct([2, 5]), 10)
        
        # Case B: 2 (0010) and 6 (0110) -> share the 2nd bit, no valid pair -> 0
        self.assertEqual(self.sol.maxProduct([2, 6]), 0)

    def test_duplicate_values(self):
        """
        Tests when duplicate values are present. Identical numbers (except 0, which isn't in constraints)
        always share set bits with themselves, so they cannot form a valid pair together.
        """
        # [4, 4] -> 4 (0100) and 4 (0100) share bits -> 0
        self.assertEqual(self.sol.maxProduct([4, 4]), 0)
        
        # [4, 4, 3] -> (4, 3) is valid -> 12
        self.assertEqual(self.sol.maxProduct([4, 4, 3]), 12)

    def test_maximum_value_constraints(self):
        """
        Tests upper bound values for elements (up to 10^6).
        1000000 in binary is 11110100001001000000 (20 bits).
        We pair it with a number that fills the remaining/opposite bit positions.
        """
        # 1000000 = 0xF4240
        # Complement mask within bounds, e.g., 9 (1001) -> 0x9. No shared bits.
        # 0xF4240 & 0x9 == 0
        nums = [1000000, 9]
        self.assertEqual(self.sol.maxProduct(nums), 9000000)

    def test_large_input_performance_and_correctness(self):
        """
        Tests a larger array to ensure scalability and that the global maximum 
        is prioritized over smaller valid products.
        """
        # 1024 (2^10) and 2048 (2^11) have no common bits and yield a high product.
        # Other elements are smaller or share bits.
        nums = [1024, 2048, 7, 15, 31, 63, 127]
        self.assertEqual(self.sol.maxProduct(nums), 1024 * 2048)


if __name__ == "__main__":
    unittest.main()

File: generated_code/cpp/test_unittest_3915.cpp
#include <gtest/gtest.h>
#include <vector>
#include "solution_proxy.h"

// Test Case 1: Standard example where a valid pair exists with no overlapping bits.
// e.g., nums = [3, 4, 8, 12] -> 3 (0011) and 4 (0100) do not share bits. Product = 12.
// 3 (0011) and 8 (1000) do not share bits. Product = 24.
// 4 (0100) and 8 (1000) do not share bits. Product = 32.
// 12 (1100) and 3 (0011) do not share bits. Product = 36. Max product is 36.
TEST(MaxProductTest, StandardValidPairs) {
    Solution sol;
    std::vector<int> nums = {3, 4, 8, 12};
    EXPECT_EQ(sol.maxProduct(nums), 36LL);
}

// Test Case 2: Case where no valid pair exists because all elements share at least one bit.
// e.g., nums = [7, 15, 31] -> All have the lowest 3 bits set (111), so they all share common bits.
TEST(MaxProductTest, NoValidPairsExist) {
    Solution sol;
    std::vector<int> nums = {7, 15, 31};
    EXPECT_EQ(sol.maxProduct(nums), 0LL);
}

// Test Case 3: Minimum constraints for array length (nums.length = 2).
TEST(MaxProductTest, MinimumArrayLength) {
    Solution sol;
    
    // Case 3a: The two elements do not share bits.
    std::vector<int> nums1 = {5, 10}; // 5 = 0101, 10 = 1010
    EXPECT_EQ(sol.maxProduct(nums1), 50LL);

    // Case 3b: The two elements share bits.
    std::vector<int> nums2 = {5, 7}; // 5 = 101, 7 = 111
    EXPECT_EQ(sol.maxProduct(nums2), 0LL);
}

// Test Case 4: Array contains duplicate values.
// Duplicates inherently share the exact same set bits (unless the value is 0, but constraints say nums[i] >= 1).
// It should look for distinct indices with non-overlapping bits.
TEST(MaxProductTest, DuplicateElements) {
    Solution sol;
    std::vector<int> nums = {5, 5, 10, 10}; // 5 (0101) and 10 (1010)
    EXPECT_EQ(sol.maxProduct(nums), 50LL);
}

// Test Case 5: Elements with maximum possible value constraint (nums[i] = 10^6).
// 10^6 fits in a standard integer, but the product can be up to 10^12, requiring long long.
// 1,000,000 = 11110100001001000000 (binary)
// We pair it with an element that has non-overlapping bits, e.g., 57 = 00001011101101111111 (inverted mask within range)
TEST(MaxProductTest, MaximumValueConstraints) {
    Solution sol;
    // 1000000 is 0xF4240. An inversion within 20 bits is 0x0BDBF = 48575.
    // 1000000 * 48575 = 48,575,000,000 which fits in long long but overflows 32-bit int.
    std::vector<int> nums = {1000000, 48575};
    EXPECT_EQ(sol.maxProduct(nums), 48575000000LL);
}

// Test Case 6: Multiple pairs are valid, ensuring the algorithm picks the absolute maximum product.
TEST(MaxProductTest, SelectsAbsoluteMaximumProduct) {
    Solution sol;
    // Pairs: 
    // (1, 2) -> product 2 (0001 and 0010)
    // (16, 15) -> product 240 (10000 and 01111)
    std::vector<int> nums = {1, 2, 15, 16};
    EXPECT_EQ(sol.maxProduct(nums), 240LL);
}